rpcs3/Emu/RSX/Overlays/overlay_manager.cpp

```diff
@@ -0,0 +1,245 @@
+			}
+
+			return false;
+		}
+
+		void display_manager::cleanup_internal()
+		{
+			for (const auto& uid : m_uids_to_remove)
+			{
+				remove_uid(uid);
+			}
+
+			for (const auto& type_id : m_type_ids_to_remove)
+			{
+				remove_type(type_id);
+			}
+
+			m_uids_to_remove.clear();
+			m_type_ids_to_remove.clear();
+		}
+
+		void display_manager::on_overlay_activated(const std::shared_ptr<overlay>& item)
+		{
+			if (auto iface = std::dynamic_pointer_cast<user_interface>(item))
+			{
+				// Kick input thread if not enabled. Expect the interface to attach shortly
+				std::lock_guard lock(m_input_thread_lock);
+
+				if (!m_input_thread)
+				{
+					m_input_thread = std::make_shared<named_thread<overlay_input_thread>>();
+					(*m_input_thread)([this]()
+					{
+						input_thread_loop();
+					});
+				}
+			}
+		}
+
+		void display_manager::attach_thread_input(
+			u32 uid,
+			std::function<void()> on_input_loop_enter,
+			std::function<void(s32)> on_input_loop_exit,
+			std::function<s32()> input_loop_override)
+		{
+			if (auto iface = std::dynamic_pointer_cast<user_interface>(get(uid)))
+			{
+				std::lock_guard lock(m_input_thread_lock);
+				m_input_token_stack.emplace_front(
+					std::move(iface),
+					on_input_loop_enter,
+					on_input_loop_exit,
+					input_loop_override);
+			}
+		}
+
+		void display_manager::on_overlay_removed(const std::shared_ptr<overlay>& item)
+		{
+			if (!dynamic_cast<user_interface*>(item.get()))
+			{
+				// Not instance of UI, ignore
+				return;
+			}
+
+			std::lock_guard lock(m_input_thread_lock);
+			for (auto& entry : m_input_token_stack)
+			{
+				if (entry.target->uid == item->uid)
+				{
+					// Release
+					entry.target = {};
+					break;
+				}
+			}
+
+			// The top must never be an empty ref. Pop all empties.
+			while (!m_input_token_stack.empty() && !m_input_token_stack.front().target)
+			{
+				m_input_token_stack.pop_front();
+			}
+		}
+
+		void display_manager::input_thread_loop()
+		{
+			while (!m_input_thread_abort)
+			{
+				input_thread_context_t input_context;
+				{
+					reader_lock lock(m_input_thread_lock);
+					if (!m_input_token_stack.empty())
+					{
+						input_context = m_input_token_stack.front();
+						m_input_token_stack.pop_front();
+					}
+				}
+
+				if (input_context.target)
+				{
+					if (input_context.input_loop_prologue)
+					{
+						input_context.input_loop_prologue();
+					}
+
+					s32 result = 0;
+					if (!input_context.input_loop_override) [[ likely ]]
+					{
+						result = input_context.target->run_input_loop();
+					}
+					else
+					{
+						result = input_context.input_loop_override();
+					}
+
+					if (input_context.input_loop_epilogue)
+					{
+						input_context.input_loop_epilogue(result);
+					}
+					else if (result && result != user_interface::selection_code::canceled)
+					{
+						rsx_log.error("Input loop exited with error code=%d", result);
+					}
+				}
+				else
+				{
+					thread_ctrl::wait_for(1000);

```

Comment:
I'd rather have this up top as early continue.
Also, in case this thread stays active with closed overlays, I'd rather have this sleep increased during emu pause (I didn't read that deep into the manager shenanigans, so no idea if that's the case)

Reply:
Maybe you can let it sleep continuously and wake it up in attach_thread_input

Reply:
Changed to use lf_queue + pop_all_reversed.